frontend/frontend.go

```diff
@@ -122,3 +130,36 @@ func streamAssignments(ctx context.Context, assignments chan *pb.Assignment, err
 		assignments <- resp.Assignment
 	}
 }
+
+func connectFrontendServer() (*grpc.ClientConn, error) {
+	var err error
+	var conn *grpc.ClientConn
+	if len(os.Args) > 1 && os.Args[1][:5] == "-test" {
+		var l net.Listener
+		l, err = net.Listen("tcp", "localhost:0")
+		if err != nil {
+			panic(err)
+		}
+		conn, err = grpc.Dial(l.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			return nil, fmt.Errorf("error dialiing to mock: %w", err)
+		}
+		gsrv := grpc.NewServer()
+		ff := omtest.FakeFrontend{}
+		pb.RegisterFrontendServiceServer(gsrv, &ff)
+
+		// Run grpc mock server
+		go func() {
+			log.Println("Mock server start:", l.Addr())
+			if err = gsrv.Serve(l); err != nil {
+				panic(err)
+			}
+		}()
+	} else {
+		conn, err = grpc.Dial(frontendAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			return nil, fmt.Errorf("error dialing open match: %w", err)
+		}
+	}
+	return conn, nil
+}
```

Comment:
Sorry if my opinion is not go style since I'm new to golang. 

Could you spin up gRPC fake server before the test start rather than depending on `os.Args` ? 

You can add the arguments of `streamAssignments` and `connectFrontendServer` if needed.

Reply:
I'd like to confirm your opinit.

Do you think about that I add the argument such like testMode to frontend and switch processing based on argument value?

Following is my imagine.
```
var conn *grpc.ClientConn

func connectFrontendServer(testMode) (*grpc.ClientConn, error) {
 if testMode == false {
     // connect to the actual open-match frontend
     conn, err = grpc.Dial(frontendAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
 } else if testMode == true {
    // connect to the fake frontend
    conn, err = grpc.Dial(mockServerAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
}
```

Is my understanding correct?

Reply:
@govargo 

Thanks for your comment. My idea is like below for example. 

```go
func streamAssignments(ctx context.Context, assignments chan *pb.Assignment, errs chan error) {
        url := os.Getenv("FRONTEND_URL")
	conn, err := connectFrontendServer(url)
	if err != nil {
		errs <- err
	}
	defer conn.Close()
	// some work
}

func connectFrontendServer(url) (*grpc.ClientConn, error) {
     conn, err = grpc.Dial(url, grpc.WithTransportCredentials(insecure.NewCredentials()))
     ...
}
```

or like this

```go

url := (get url somhow)

func streamAssignments(ctx context.Context, assignments chan *pb.Assignment, errs chan error, url string) {
	conn, err := connectFrontendServer(url)
	if err != nil {
		errs <- err
	}
	defer conn.Close()
	// some work
}

func connectFrontendServer(url) (*grpc.ClientConn, error) {
     conn, err = grpc.Dial(url, grpc.WithTransportCredentials(insecure.NewCredentials()))
     ...
}
```

By using `TestMain`, you can set up the server for the test.